### DISCUSSION TITLE
how to add permission to access_token or userinfo

### DISCUSSION THREAD

**Discussion by Basil1991** [keycloak/keycloak#20322] [Q&A]
i want to use only authentication, i need add both role and permission code in access_token or userinfo. what should i do?

**Comment by souravs17031999** [comment]
@Basil1991 what do you mean you need both role and permission code ? Does that mean you need exact roles and permissions mapped for users in the token that you receive while logging in ?

You can check out client-scope protocol mappers which allows you to add custom claims to your tokens.
https://github.com/mschwartau/keycloak-custom-protocol-mapper-example/blob/master/README.md

**Reply by Basil1991** [reply]
@souravs17031999 in my opnion，role like ‘user’，‘admin’，permission like ‘user_manager’,'order_create','order_query'.
without enable authorization in keycloak  how can i use permission  concepts.
in my case ,i want to return token or userinfo like {roles:"xx",permission_code:"xxx"}, that application use roles and permission do what they want to do(like use shiro,spring security).

**Comment by sridhar1495** [comment]
@Basil1991 There is one thing you need to note is that in KeyCloak the authentication is happen at the realm level. So that you will be getting the roles which are given for the user in the realm roles. But the permission you will be configured in the Client level by enabling the Authorization. 
Based on this the KeyCloak will give you the permissions during the Authorization part only not at the authentication part because you have to provide the Client details along with the Authentication token. You can get the details in the following link
https://www.keycloak.org/docs/latest/authorization_services/index.html#_service_obtaining_permissions

As far as I know you can't get the permission details during the Authentication.

**Reply by Basil1991** [reply]
@sridhar1495 thanks.  so i should use just realm level authentication , then put client level permission  in the application.

**Reply by sridhar1495** [reply]
@Basil1991 yes you are correct. As per the document link I given itself they have mentioned to get the permission we need to follow those steps. I had gone through different forum and looked for alternatives but could not find any solution for getting the permission in authentication step. 
Also, I have discussed with my architect about this, and they are ok with this approach, as per the security standards also it is fine to go with this approach since it will comes under the defense in depth.

**Comment by Basil1991** [comment]
@sridhar1495 thanks~
